### PR TITLE
Numba formal integral

### PR DESCRIPTION
**Pull request by Rodot-**
**Description**

Implementation of the formal integral with numba.  New tests added for the numba version of various helper functions used to compute the formal integral.

**Motivation and context**

We would like to move TARDIS fully away from C.  The formal integral was the last component of TARDIS written entirely in C.

**How has this been tested?**

New tests have been added for the numba implementation.  Full runs that compute formal integrals are compared to reference-data generated using the C formal integral.  All tests pass.

- [x] Testing pipeline.
- [ ] Other.

**Type of change**

- [ ] Bug fix. 
- [ ] New feature. 
- [ ] Breaking change. 
- [x] None of the above. 

Refactor from C to Python

**Checklist**

- [ ] My change requires a change to the documentation.
    - [ ] I have updated the documentation accordingly.
    - [ ] (optional) I have built the documentation on my fork following [the instructions](https://tardis-sn.github.io/tardis/development/documentation_preview.html).
- [x] I have assigned and requested two reviewers for this pull request.

### CODE REVIEW

**Review by andrewfullard — APPROVED** [review]

**Comment by Rodot-** [comment]
One thing to note: I removed tardis/montecarlo/tests/test_montecarlo.py since it was mostly C tests, but when I merged I noticed there had been some updates to it.  Should this file be added back in and modified to only do the numba side of things or is it fine being removed?  @andrewfullard

**Comment by Rodot-** [comment]
Okay, well, I added them back in just in case, they all skip anyway and I modified them not to include any C code references.  But I left them just in case we want to move some of these tests over to a different file later so the work isn't lost.

I think this is ready to go

File: tardis/montecarlo/montecarlo_numba/formal_integral.py [hunk 1]
```diff
@@ -1,53 +1,236 @@
+
+import sys
+
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/formal_integral.py:3]:
```suggestion
```

File: tardis/montecarlo/montecarlo_numba/formal_integral.py [hunk 2]
```diff
@@ -1,53 +1,236 @@
...
 from astropy import units as u
 from tardis import constants as const
-from numba import njit
+from numba import njit, char, float64, int64, jitclass, typeof, byte, prange
 import pdb
 
+from tardis.montecarlo.montecarlo_numba.numba_config import SIGMA_THOMSON
+from tardis.montecarlo.montecarlo_numba import njit_dict, njit_dict_no_parallel
+from tardis.montecarlo.montecarlo_numba.numba_interface \
+    import numba_plasma_initialize, NumbaModel, NumbaPlasma
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/formal_integral.py:17]:
```suggestion
from tardis.montecarlo.montecarlo_numba.numba_interface (
    import numba_plasma_initialize, NumbaModel, NumbaPlasma)
```

File: tardis/montecarlo/montecarlo_numba/formal_integral.py [hunk 3]
```diff
@@ -1,53 +1,236 @@
...
-from tardis.montecarlo.montecarlo import formal_integral
 from tardis.montecarlo.spectrum import TARDISSpectrum
 
 C_INV = 3.33564e-11
 M_PI = np.arccos(-1)
 KB_CGS = 1.3806488e-16
 H_CGS = 6.62606957e-27
-SIGMA_THOMSON = 6.652486e-25
-
+#SIGMA_THOMSON = 6.652486e-25
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/formal_integral.py:25]:
```suggestion
```

File: tardis/montecarlo/montecarlo_numba/formal_integral.py [hunk 4]
```diff
@@ -1,53 +1,236 @@
...
+    ('plasma', NumbaPlasma.class_type.instance_type),
+    ('points', int64)
+]
+@jitclass(integrator_spec)
+class NumbaFormalIntegrator(object):
+    '''
+    Helper class for performing the formal integral
+    with numba.
+    '''
+    #def __init__(self, model, plasma, runner, points=1000):
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/formal_integral.py:186]:
```suggestion
```

File: tardis/montecarlo/montecarlo_numba/formal_integral.py [hunk 5]
```diff
@@ -616,26 +642,30 @@ def binary_search(x, x_insert, imin, imax, result):
...
-    for i in range(N):
-        opp[i] = R_max / (N - 1) * (i)
-    return opp
+@njit(**njit_dict_no_parallel)
+def calculate_p_values(R_max, N):
+    '''This can probably be replaced with a simpler function'''
+    return np.arange(N).astype(np.float64) * R_max / (N - 1)
+    #for i in range(N):
+    #    opp[i] = R_max / (N - 1) * (i)
+    #return opp
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/formal_integral.py:671]:
```suggestion
```

**Review by wkerzendorf — APPROVED** [review]

**Pull request merged**